chia/wallet/wallet_singleton_store.py

```diff
@@ -0,0 +1,281 @@
+
+from chia.types.blockchain_format.coin import Coin
+from chia.types.blockchain_format.program import Program
+from chia.types.blockchain_format.sized_bytes import bytes32
+from chia.util.db_wrapper import DBWrapper2, execute_fetchone
+from chia.util.ints import uint32, uint64
+from chia.wallet.lineage_proof import LineageProof
+from chia.wallet.singleton_coin_record import SingletonCoinRecord
+
+_T_WalletSingletonStore = TypeVar("_T_WalletSingletonStore", bound="WalletSingletonStore")
+
+
+def _row_to_singleton_record(row: Row, custom_data: Optional[Dict[str, Any]] = None) -> SingletonCoinRecord:
+    return SingletonCoinRecord(
+        coin=Coin(row[1], row[2], uint64(row[3])),
+        singleton_id=bytes32(row[4]),
+        wallet_id=uint32(row[5]),
+        inner_puzzle=Program.from_bytes(row[6]),
+        inner_puzzle_hash=bytes32(row[7]),
+        confirmed=bool(row[8]),
+        confirmed_at_height=uint32(row[9]),
+        spent_height=uint32(row[10]),
+        lineage_proof=LineageProof.from_bytes(row[11]),
+        custom_data=custom_data,
+        generation=uint32(row[12]),
+        timestamp=uint64(row[13]),
+    )
+
+
+class WalletSingletonStore:
+    """
+    WalletSingletonStore keeps track of all user created singletons and necessary wallet data
+
+    Coins can be added as 'confirmed=False' when they are created in a wallet transaction. Once the
+    transaction is confirmed, the wallet should update to 'confirmed=True'.
+
+    The custom_data field in SingletonCoinRecord is split into the singleton_data table, with a
+    Foreign Key on coin_id.
+
+    NOTE: For singleton_data to correctly impose foreign key constraints, the DBWrapper must be
+    created with foreign_keys=True
+    """
+
+    db_wrapper: DBWrapper2
+
+    @classmethod
+    async def create(cls: Type[_T_WalletSingletonStore], wrapper: DBWrapper2) -> _T_WalletSingletonStore:
+        self = cls()
+        self.db_wrapper = wrapper
+
+        async with self.db_wrapper.writer_maybe_transaction() as conn:
+            await conn.execute(
+                (
+                    "CREATE TABLE IF NOT EXISTS singletons("
+                    "coin_id blob PRIMARY KEY,"
+                    " parent_coin_info text,"
+                    " puzzle_hash text,"
+                    " amount blob,"
+                    " singleton_id blob,"
+                    " wallet_id int,"
+                    " inner_puzzle blob,"
+                    " inner_puzzle_hash blob,"
+                    " confirmed tinyint,"
+                    " confirmed_at_height int,"
+                    " spent_height int,"
+                    " lineage_proof blob,"
+                    " generation int,"
+                    " timestamp int)"
+                )
+            )
+
+            await conn.execute("CREATE INDEX IF NOT EXISTS singleton_id_index ON singletons(singleton_id)")
+            await conn.execute("CREATE INDEX IF NOT EXISTS wallet_id_index ON singletons(wallet_id)")
```

Comment:
I think you need an index for `generation` as well. You filter and order by it.

Reply:
Thanks!